### PR TITLE
EncodingError/DecodingError bridging fixes and conveniences

### PR DESCRIPTION
**Pull request by itaiferber**
**What's in this pull request?**
As laid out in a [proposed update to SE-0166 and SE-0167](https://lists.swift.org/pipermail/swift-evolution/Week-of-Mon-20170619/037672.html), there are `EncodingError`/`DecodingError` updates that we wanted to introduce to the Codable API. This implements the remainder of those updates which had not yet been implemented:

* `underlyingError` exposition on `EncodingError` and `DecodingError`
* `.dataCorrupted` conveniences for user code
* Fixing error bridging to `NSError`

This subsumes #10674.

### CODE REVIEW

**Comment by itaiferber** [comment]
This is the correct version which should work on Linux as well.

**Comment by itaiferber** [comment]
@swift-ci Please test

File: stdlib/public/core/Codable.swift [hunk 1]
```diff
@@ -2026,20 +2034,63 @@ public enum EncodingError : Error {
...
+                "NSDebugDescription": context.debugDescription
+            ]
+
+            if let underlyingError = context.underlyingError {
+                userInfo["NSUnderlyingError"] = underlyingError
+            }
+
+            return userInfo as AnyObject
+        #else
+            return nil
```
Comment (phausler) [inline stdlib/public/core/Codable.swift]: can we have a callback here to hook a return for swift-corelibs-foundation?

**Review by phausler — APPROVED** [review]
Overall looks good to me

**Pull request merged**